lib/scnnr/client.rb

```diff
@@ -20,66 +20,121 @@ def config
       uri = construct_uri('coordinates', %i[target], options)
       payload = {
         item: { category: category, labels: labels },
         taste: TASTES.each_with_object({}) { |key, memo| memo[key] = taste[key] if taste[key] },
       }
-      response = post_connection(uri, options).send_json(payload)
+      response = post(uri, payload, options)
       Response.new(response).build_coordinate
     end
 
     private
 
-    def merge_options(options = {})
-      self.config.to_h.merge(options)
+    def recognize(options, &block)
+      options = default_options.merge options
+      return block.call(options) if without_timeout?(options[:timeout])
+
+      validate_timeout(options[:timeout])
+      timeout_at = Scnnr::PollingUtils.timeout_at(options[:timeout])
+
+      result = recognize_start(timeout_at, options) do |opts|
+        block.call(opts)
+      end
+
+      return result unless result == :poll
+
+      recognize_poll(result.id, timeout_at, options)
+    end
+
+    def recognize_start(timeout_at, options = {}, &block)
+      Scnnr::PollingUtils.start timeout_at do
+        timeout = Scnnr::PollingUtils.calculate_timeout(timeout_at)
+        block.call(options.merge({ timeout: timeout }))
+      end
+    end
+
+    def recognize_poll(id, timeout_at, options = {})
+      Scnnr::PollingUtils.poll timeout_at do
+        timeout = Scnnr::PollingUtils.calculate_timeout(timeout_at)
+        get_recognition(id, options.merge({ timeout: timeout }))
+      end
+    end
+
+    def without_timeout?(timeout)
+      timeout.nil? || timeout.zero?
+    end
+
+    def validate_timeout(timeout)
+      return if without_timeout?(timeout)
+      return if timeout.is_a?(Integer) && timeout.positive?
+      return if timeout.is_a?(Float::INFINITY)
+
+      raise ArgumentError, "timeout must be Integer or Float::INFINITY, but given: #{timeout}"
+    end
+
+    def get_recognition(recognition_id, options = {})
+      uri = construct_uri("recognitions/#{recognition_id}", %i[timeout], options)
+      response = get(uri, options).send_request_with_retries
+      Response.new(response).build_recognition
+    end
+
+    def recognize_image_with_file(file, options = {})
+      uri = construct_uri('recognitions', %i[timeout public], options)
+      response = post_file(uri, file, options)
+      Response.new(response).build_recognition
+    end
+
+    def recognize_image_with_url(url, options = {})
+      uri = construct_uri('remote/recognitions', %i[timeout force], options)
+      response = post(uri, { url: url }, options)
+      Response.new(response).build_recognition
+    end
+
+    def default_options
+      self.config.to_h
     end
 
     def construct_uri(path, allowed_params, options = {})
-      Routing.new(
+      RoutingUtils.to_url(
         path, options[:api_version],
         options, allowed_params
-      ).to_url
+      )
     end
 
-    def get_connection(uri, options = {})
+    def get(uri, options = {})
```

Comment:
What is the reason why we need to rename this method?

Reply:
Just make it looks simple, but just a proposal, it's not saying the old is bad, please do not misunderstand me.